### PR TITLE
Add new native module template support to react-native-windows-init

### LINKED ISSUE #3201: Enable the CLI to create a C# native module library that consumes Microsoft.ReactNative
Project should be set up similar to the SampleLibraryCS sample:

- [ ] C# component library
- [ ] Depends only on Microsoft.ReactNative and Microsoft.ReactNative.Managed
- [ ] Provides a ReactPackageProvider that automatically includes all native modules, and view managers
- [ ] Provides a simple sample native module

### LINKED ISSUE #3203: Enable the CLI to create a C++/WinRT native module library that consumes Microsoft.ReactNative
Project should be set up similar to the SampleLibraryCPP sample:

- [ ] C++/WinRT component library
- [ ] Depends only on Microsoft.ReactNative and Microsoft.ReactNative.Cxx
- [ ] Provides a ReactPackageProvider that automatically includes all native modules, and view managers
- [ ] Provides a simple sample native module

### PR DESCRIPTION
**Pull request by jonthysell**
This PR extends `react-native-windows-init` with a `--projectType [app|lib]` option and a set of new native module lib templates for both C++ and C#. This provides a much needed mechanism for developers wishing to add windows implementation to an existing native module (esp. community modules) without having to set up the VS files manually.

Closes #3203 
Closes #3201

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/5921)

### CODE REVIEW

File: packages/@react-native-windows/cli/src/generator-windows/index.ts [hunk 1]
```diff
@@ -335,48 +376,64 @@ export async function copyProjectTemplateAndReplace(
...
+  if (fs.existsSync(path.join(sharedPath, projDir))) {
+    // Once we are publishing to nuget.org, this shouldn't be needed anymore
+    if (options.experimentalNuGetDependency) {
+      [
+        {
+          from: path.join(sharedPath, projDir, 'NuGet.Config'),
+          to: path.join(windowsDir, 'NuGet.Config'),
+        },
+      ].forEach(mapping =>
+        copyAndReplaceWithChangedCallback(
```
[the remark this answers is not in the document]
Comment (NickGerleman) [reply packages/@react-native-windows/cli/src/generator-windows/index.ts:389]: I'm going to see if I can find an eslint rule for unhandled promises

[the remark this answers is not in the document]
Comment (NickGerleman) [reply packages/@react-native-windows/cli/src/generator-windows/index.ts:389]: Got that working and it found some other cases. I'll fix those up in a separate PR, so don't worry about it here.

File: vnext/template/cpp-lib/proj/MyLib.vcxproj [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+      <PrecompiledHeaderFile>pch.h</PrecompiledHeaderFile>
+      <PrecompiledHeaderOutputFile>$(IntDir)pch.pch</PrecompiledHeaderOutputFile>
+      <WarningLevel>Level4</WarningLevel>
+      <AdditionalOptions>%(AdditionalOptions) /bigobj</AdditionalOptions>
+      <DisableSpecificWarnings>4453;28204</DisableSpecificWarnings>
+      <PreprocessorDefinitions>_WINRT_DLL;%(PreprocessorDefinitions)</PreprocessorDefinitions>
+      <AdditionalUsingDirectories>$(WindowsSDK_WindowsMetadata);$(AdditionalUsingDirectories)</AdditionalUsingDirectories>
+    </ClCompile>
+    <Link>
+      <SubSystem>Console</SubSystem>
```
Comment (asklar) [inline vnext/template/cpp-lib/proj/MyLib.vcxproj:108]: this seems surprising?

File: vnext/template/cpp-lib/proj/MyLib.vcxproj [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+    <None Include="packages.config" />
+    <None Include="PropertySheet.props" />
+  </ItemGroup>
+  <Import Project="$(VCTargetsPath)\Microsoft.Cpp.targets" />
+  <ImportGroup Label="ReactNativeWindowsTargets">
+    <Import Project="$(ReactNativeWindowsDir)\PropertySheets\External\Microsoft.ReactNative.Uwp.CppLib.targets" Condition="Exists('$(ReactNativeWindowsDir)\PropertySheets\External\Microsoft.ReactNative.Uwp.CppLib.targets')" />
+  </ImportGroup>
+  <Target Name="EnsureReactNativeWindowsTargets" BeforeTargets="PrepareForBuild">
+    <PropertyGroup>
+      <ErrorText>This project references targets in your node_modules\react-native-windows folder. The missing file is {0}.</ErrorText>
```
Comment (asklar) [inline vnext/template/cpp-lib/proj/MyLib.vcxproj:152]: missing "that are missing"

File: vnext/template/cpp-lib/src/ReactNativeModule.h [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+#include "JSValue.h"
+#include "NativeModules.h"
+
+// clang-format off
+
+using namespace winrt::Microsoft::ReactNative;
+
+namespace winrt::{{ namespaceCpp }}
+{
+
```
Comment (asklar) [inline vnext/template/cpp-lib/src/ReactNativeModule.h:12]: add a comment with a link to an aka.ms pointing to the documentation for writing native modules in C++ (same for C# lib)

File: vnext/template/cpp-lib/src/pch.h [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+#include <winrt/{{ xamlNamespace }}.Navigation.h>
+#include <winrt/{{ xamlNamespace }}.h>
+
+#include <winrt/Microsoft.ReactNative.h>
+
+#include <winrt/Microsoft.UI.Xaml.Automation.Peers.h>
+#include <winrt/Microsoft.UI.Xaml.Controls.Primitives.h>
+#include <winrt/Microsoft.UI.Xaml.Controls.h>
+#include <winrt/Microsoft.UI.Xaml.Media.h>
+#include <winrt/Microsoft.UI.Xaml.XamlTypeInfo.h>
```
Comment (asklar) [inline vnext/template/cpp-lib/src/pch.h:25]: nice :)

File: vnext/template/cs-lib/proj/MyLib.csproj [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+  <Import Project="$(MSBuildExtensionsPath)\Microsoft\WindowsXaml\v$(VisualStudioVersion)\Microsoft.Windows.UI.Xaml.CSharp.targets" />
+  <ImportGroup Label="ReactNativeWindowsPropertySheets">
+    <Import Project="$(ReactNativeWindowsDir)\PropertySheets\External\Microsoft.ReactNative.Uwp.CSharpLib.props" Condition="Exists('$(ReactNativeWindowsDir)\PropertySheets\External\Microsoft.ReactNative.Uwp.CSharpLib.props')" />
+  </ImportGroup>
+  <ImportGroup Label="ReactNativeWindowsTargets">
+    <Import Project="$(ReactNativeWindowsDir)\PropertySheets\External\Microsoft.ReactNative.Uwp.CSharpLib.targets" Condition="Exists('$(ReactNativeWindowsDir)\PropertySheets\External\Microsoft.ReactNative.Uwp.CSharpLib.targets')" />
+  </ImportGroup>
+  <Target Name="EnsureReactNativeWindowsTargets" BeforeTargets="PrepareForBuild">
+    <PropertyGroup>
+      <ErrorText>This project references targets in your node_modules\react-native-windows folder. The missing file is {0}.</ErrorText>
```
Comment (asklar) [inline vnext/template/cs-lib/proj/MyLib.csproj:147]: missing missing too

**Review by asklar — APPROVED** [review]

**Pull request merged**